File: Data Structures & Algorithms/redundant-connection/submission-1.py
class Solution:
    def findRedundantConnection(self, edges: List[List[int]]) -> List[int]:
        n = len(edges)
        parent = [i for i in range(n+1)]
        def find(x):
            if parent[x] != x:
                parent[x] = find(parent[x])

            return parent[x]
        
        def union(x, y):
            rootX = find(x)
            rootY = find(y)
            if rootX == rootY:
                return False 
            
            parent[rootX] = rootY 
            return True
        
        for u, v in edges:
            if not union(u, v):
                return [u, v]
        
        return []

File: Data Structures & Algorithms/count-connected-components/submission-1.java
class Solution {
    private int[] parent;
    private int[] rank;
    private int count;

    public int countComponents(int n, int[][] edges) {
        count = n;
        parent = new int[n];
        rank = new int[n];
        for (int i = 0; i < n; i++) {
            parent[i] = i;
        }

        
        for (int[] edge : edges) {
            int u = edge[0];
            int v = edge[1];

            union(u, v);
        }
        return count;
    }

    private int find(int x) {
        if (parent[x] != x) {
            parent[x] = find(parent[x]);
        }
        return parent[x];
    }

    private void union(int x, int y) {
        int rootX = find(x);
        int rootY = find(y);
        if (rootX == rootY) return;
        if (rank[rootX] < rank[rootY]) {
            int tmp = rootX;
            rootX = rootY;
            rootY = tmp;
        }
        parent[rootY] = rootX;
        rank[rootX] += rank[rootY];
        count -= 1;
    }
}
